### PR TITLE
Implement Books API Integration

### PR DESCRIPTION
**Pull request by eliskavo**
- Implement API logic
- Clean up code structure and error handling
- Create reusable Loading component with animation
- Add pagination for trending books

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -35,7 +35,7 @@
   "rules": {
     "strict": [2, "never"], // Avoid any unsafe functions that would not be allowed under strict mode
     "default-case": 0, // Allow switch without a default case
-    "no-console": 2, // Forbid console logs
+    "no-console": "off", // Forbid console logs
```
Comment (GLObus303) [inline .eslintrc]: Let's not do this :). 

a) we use numbers 0, 1, 2 so this only brings inconsistency
b) allow only console.errors https://eslint.org/docs/latest/rules/no-console

File: types/types.ts [hunk 2]
```diff
@@ -20,3 +21,20 @@ export type Review = {
   reviewDate: string;
   bookId: number;
 };
+
+export type LoadingProps = {
+  message?: string;
+};
+
+export type FetchBooksParams = {
```
Comment (GLObus303) [inline types/types.ts]: There is no need to separate this (even Vite might complain). On the other hand, colocation(having it next to the function) is the place where everybody would search for it.

Especially the separation for the props type would make the project veeeery messy very fast 💩

File: src/utils/parseId.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export const parseBookId = (id: string) => id.split('/')[2];
+
+export const parseAuthorId = (id: string) => id.split('/')[2];
```
Comment (GLObus303) [inline src/utils/parseId.ts]: Even sometimes I could get behind duplication like this, I believe we don't need to do it in the same file next to each other :blush: . Change the name of this to something more generic `parseItemIdFromUri` and reuse it for both

File: types/types.ts [hunk 4]
```diff
@@ -1,16 +1,17 @@
-export interface Book {
-  id: number;
+export type Book = {
```
Comment (GLObus303) [inline types/types.ts]: Create a folder "model" in src. And add `Book.ts` there with this type. It is Java approach a tad 😬 , but I personally like it.

It gonna look later something like this:

CleanShot 2024-12-08 at 22 43 43

Comment (GLObus303) [reply types/types.ts]: For the size of our project is perfectly fine 💪 . The problem 🙈  with what you have here is that it would very quickly grow and the "model" layer has no disadvantage I believe, so let's use it right away 🚀

File: src/pages/Books/Books.tsx [hunk 5]
```diff
@@ -1,51 +1,93 @@
...
+  const [currentPage, setCurrentPage] = useState(1);
+  const [totalBooks, setTotalBooks] = useState(0);
 
-  const filteredBooks = books.filter(
-    (book) =>
-      book.title.toLowerCase().includes(searchQuery.toLowerCase()) ||
-      book.author.toLowerCase().includes(searchQuery.toLowerCase()),
-  );
+  const booksPerPage = 30;
+
```
Comment (GLObus303) [inline src/pages/Books/Books.tsx:17]: Move constants like this above the component, like you have it here

CleanShot 2024-12-08 at 22 46 50


I personally like naming it all caps snake case, so BOOKS_PER_PAGE. I am not sure if it is necessarily best practice but it nicely visually differentiate between some random variable and really-not-changing constant

File: src/pages/Books/Books.tsx [hunk 6]
```diff
@@ -1,51 +1,93 @@
...
     <Layout>
       <div className={style.bookSection}>
-        <h1 className={style.currentReadingTitle}>we are currently reading</h1>
-        {currentlyReadingBook && <CurrentBookCard books={books} />}
+        {currentlyReadingBook && (
+          <div>
+            <h2 className={style.currentReadingTitle}>
+              We are currently reading:
+            </h2>
+            <CurrentBookCard book={currentlyReadingBook} />
```
Comment (GLObus303) [inline src/pages/Books/Books.tsx]: CleanShot 2024-12-08 at 22 53 10

File: src/components/Loading/Loading.module.scss [hunk 7]
```diff
@@ -0,0 +1,67 @@
+// src/components/Loading/Loading.module.scss
```
Comment (GLObus303) [inline src/components/Loading/Loading.module.scss]: Tell chatgpt to burn these :D

File: src/components/Loading/Loading.tsx [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import search from '../../assets/images/search.png';
+import { LoadingProps } from '../../../types/types';
+import style from './Loading.module.scss';
+
+export const Loading: React.FC<LoadingProps> = ({ message }) => (
```
Comment (GLObus303) [inline src/components/Loading/Loading.tsx:8]: Sweet looking loading 🤩 

CleanShot 2024-12-08 at 23 00 39

File: src/components/Loading/Loading.module.scss [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  color: var(--text-secondary);
+  display: flex;
+  align-items: center;
+}
+
+.dots {
+  display: inline-flex;
+  margin-left: 4px;
+}
+.dot {
```
Comment (GLObus303) [inline src/components/Loading/Loading.module.scss:41]: New line above. Let's remove these inconsistencies during self code review which you definitely did at least twice 🫶

File: src/api/bookApi.tsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
+import { parseBookId } from '../utils/parseId';
+import { FetchBooksParams } from '../../types/types';
+
+const page = 1;
+const limit = 30;
+
+export const BookAPI = {
```
Comment (GLObus303) [inline src/api/bookApi.tsx]: Even this approach is perfectly fine. It reminds me of classes too much and that gives me PTSD. 

On top of it, it adds unnecessary indentation just so you can import one thing. I (personally!) find it visually cluttering for no gain. 

Export two separate function. Use arrow functions, always!

File: src/api/bookApi.tsx [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  async fetchSearchBooks({ query }: FetchBooksParams) {
+    const offset = (page - 1) * limit;
+    const url = `https://openlibrary.org/search.json?q=${query}&fields=key,title,author_name,cover_i,editions&lang=eng,cze&limit=${limit}&offset=${offset}`;
```
Comment (GLObus303) [inline src/api/bookApi.tsx]: Btw, your approach is fine. You could also do this:

```
    const searchParams = new URLSearchParams({
      q: query,
      fields: 'key,title,author_name,cover_i,editions',
      lang: 'eng,cze',
      limit: limit.toString(),
      offset: offset.toString()
    });

    const url = `https://openlibrary.org/search.json?${searchParams.toString()}`;
    
```

You decide. Whatever you prefer

File: src/api/bookApi.tsx [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+    try {
+      const response = await fetch(url);
+      const data = await response.json();
+
+      return {
+        books: data.docs.map((book: any) => ({
```
Comment (GLObus303) [inline src/api/bookApi.tsx]: Type the book. Fetch does not have generic parameter (fetch<ThisIsGenericParamater>) so you got to explicit cast it with `as SomeType`.

File: src/api/bookApi.tsx [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+      console.error('Error fetching search books:', error);
+
+      return { books: [], total: 0 };
+    }
+  },
+
+  async fetchTrendingBooks({}: FetchBooksParams) {
+    try {
+      const response = await fetch(
+        'https://openlibrary.org/trending/daily.json?limit=150',
```
Comment (GLObus303) [inline src/api/bookApi.tsx]: Why do you fetch 150 when you limit is 30? :D

File: src/api/bookApi.tsx [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        total: data.num_found,
+      };
+    } catch (error) {
+      console.error('Error fetching search books:', error);
+
+      return { books: [], total: 0 };
+    }
+  },
+
+  async fetchTrendingBooks({}: FetchBooksParams) {
```
Comment (GLObus303) [inline src/api/bookApi.tsx]: Are you sure this type should be here? 🙈

File: src/api/bookApi.tsx [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    try {
+      const response = await fetch(
+        'https://openlibrary.org/trending/daily.json?limit=150',
+      );
+      const data = await response.json();
+
+      const allBooks = data.works.map((book: any) => ({
+        id: parseBookId(book.key),
+        title: book.title || 'Untitled',
+        author: book.author_name?.[0] || 'Unknown',
```
Comment (GLObus303) [inline src/api/bookApi.tsx]: Good job you thought about the defaults! 💪

File: src/components/Loading/Loading.module.scss [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    animation-delay: 0s;
+  }
+
+  &:nth-child(2) {
+    animation-delay: 0.2s;
+  }
+
+  &:nth-child(3) {
+    animation-delay: 0.4s;
+  }
```
Comment (GLObus303) [inline src/components/Loading/Loading.module.scss:55]: 😮

File: src/components/Loading/Loading.tsx [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+  <div className={style.loadingContainer}>
+    <div className={style.loadingText}>
+      {message}
+      <span className={style.dots}>
+        <span className={style.dot}>.</span>
+        <span className={style.dot}>.</span>
+        <span className={style.dot}>.</span>
+      </span>
+    </div>
+    <img src={search} alt="Loading books" className={style.illustration} />
```
Comment (GLObus303) [inline src/components/Loading/Loading.tsx:18]: alt. GOOD


JOB 🤯

File: src/pages/BookDetail/BookDetail.module.scss [hunk 18]
```diff
@@ -100,81 +87,81 @@
...
-  @media (min-width: $screen-sm-min) {
-    font-size: 0.9rem;
-  }
-}
-
-.noReviews {
-  color: $dark-grey;
-  font-style: italic;
-}
+// .reviewsSection {
```
Comment (GLObus303) [inline src/pages/BookDetail/BookDetail.module.scss]: Purge it. No commented-out code, please. We can pull it out from the git if needed :)

File: src/components/StarsRating/StarsRating.tsx [hunk 19]
```diff
@@ -0,0 +1,14 @@
+import style from './StarsRating.module.scss';
+
+export const Stars = ({ rating }: { rating: number }) => {
+  const roundedRating = Math.round(rating * 2) / 2;
+
+  return [...Array(5)].map((_, index) => (
+    <span
+      key={index}
+      className={`${style.star} ${index < roundedRating ? style.starFilled : style.starEmpty}`}
```
Comment (GLObus303) [inline src/components/StarsRating/StarsRating.tsx]: Maybe this could be improved. Can you think of how 🤔

File: src/components/Pagination/Pagination.tsx [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+import { PaginationProps } from '../../../types/types';
+import style from './Pagination.module.scss';
+
+export const Pagination: React.FC<PaginationProps> = ({
+  currentPage,
+  totalPages,
+  onPageChange,
+  showTotalPages = true,
+}) => (
+  <div className={style.pagination}>
```
Comment (GLObus303) [inline src/components/Pagination/Pagination.tsx]: pagination could be `nav`. Check google, maybe I am bsing you

File: src/components/Pagination/Pagination.tsx [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+    </button>
+
+    <span className={style.pageInfo}>
+      Page {currentPage}
+      {showTotalPages && totalPages ? ` of ${totalPages}` : ''}
+    </span>
+
+    <button
+      onClick={() => onPageChange(true)}
+      disabled={totalPages ? currentPage >= totalPages : false}
```
Comment (GLObus303) [inline src/components/Pagination/Pagination.tsx]: I believe `!!(totalPages && currentPage >= totalPages)` could be the same

File: src/components/Pagination/Pagination.tsx [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+  totalPages,
+  onPageChange,
+  showTotalPages = true,
+}) => (
+  <div className={style.pagination}>
+    <button
+      onClick={() => onPageChange(false)}
+      disabled={currentPage === 1}
+      className={style.paginationButton}
+      aria-label="Previous page"
```
Comment (GLObus303) [inline src/components/Pagination/Pagination.tsx]: 😉

File: src/pages/Books/Books.tsx [hunk 23]
```diff
@@ -1,51 +1,93 @@
...
-      book.title.toLowerCase().includes(searchQuery.toLowerCase()) ||
-      book.author.toLowerCase().includes(searchQuery.toLowerCase()),
-  );
+  const booksPerPage = 30;
+
+  const fetchBooks = async (page: number) => {
+    setIsLoading(true);
+    try {
+      const result = searchQuery
+        ? await BookAPI.fetchSearchBooks({
```
Comment (GLObus303) [inline src/pages/Books/Books.tsx]: I wonder how it would look if you would pass the fetch function a parameter. It would be a bit more declarative - you would get rid of the branching ternary

File: src/pages/Books/Books.tsx [hunk 24]
```diff
@@ -1,51 +1,93 @@
...
+      setCurrentPage(page);
+    } catch (error) {
+      console.error('Error fetching books:', error);
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  useEffect(() => {
+    const timeoutId = setTimeout(() => {
```
Comment (GLObus303) [inline src/pages/Books/Books.tsx]: This is absolutely divine performance improvement. Lovely you have thought about it,, that by itself puts you on top of the food chain 🥇

File: src/pages/Books/Books.tsx [hunk 25]
```diff
@@ -1,51 +1,93 @@
...
+      fetchBooks(1);
+    }, 500);
+
+    return () => clearTimeout(timeoutId);
+  }, [searchQuery]);
```
Comment (GLObus303) [inline src/pages/Books/Books.tsx]: Even it might be tempting to write event handlers in effects, it might not always be the best place. -> https://react.dev/learn/you-might-not-need-an-effect#fetching-data

It also has a small bug mentioned in the paragraph above ☝️ which we are going to HAPPILLY ignore :D, cause it really doesn't matter in this case, and I am sure 80% of React developers are unaware. 

Implementing this in an event handler would be possible, but it might just be too convoluted. What you did is stellar 🌟

File: src/pages/Books/Books.tsx [hunk 26]
```diff
@@ -1,51 +1,93 @@
...
+
+  const handlePageChange = (isNext: boolean) => {
+    const nextPage = isNext ? currentPage + 1 : currentPage - 1;
+    const maxPages = Math.ceil(totalBooks / booksPerPage);
 
-  const toggleSearch = () => {
-    setIsSearchOpen(!isSearchOpen);
+    if (nextPage >= 1 && nextPage <= maxPages) {
+      fetchBooks(nextPage);
+      window.scrollTo({ top: 0, behavior: 'smooth' });
```
Comment (GLObus303) [inline src/pages/Books/Books.tsx]: Why do you want to scroll to top? Btw do you know fetchBooks is async and you scroll right away?

File: src/pages/Books/Books.tsx [hunk 27]
```diff
@@ -1,51 +1,93 @@
...
+    }
   };
```
Comment (GLObus303) [inline src/pages/Books/Books.tsx]: Should we move this to pagination? It feels somewhat better.

Comment (GLObus303) [reply src/pages/Books/Books.tsx]: What you did it great. The pagination is NOT coupled with the implementation of this page. But what if we move everything down. I believe it still will be the way EVERY pagination will work. Give it more generic names - words `books` replace with `items`.

File: src/pages/Books/Books.tsx [hunk 28]
```diff
@@ -54,14 +96,30 @@ export const Books: React.FC = () => {
...
+              <div className={style.bookGrid}>
+                {books.map((book) => (
+                  <BookCard key={book.id} book={book} />
+                ))}
+              </div>
+
+              {books.length > 0 && (
+                <Pagination
+                  currentPage={currentPage}
+                  totalPages={Math.ceil(totalBooks / booksPerPage)}
```
Comment (GLObus303) [inline src/pages/Books/Books.tsx]: This calculation should be inside I believe

File: src/pages/Books/Books.tsx [hunk 29]
```diff
@@ -1,51 +1,93 @@
...
+  const [totalBooks, setTotalBooks] = useState(0);
 
-  const filteredBooks = books.filter(
-    (book) =>
-      book.title.toLowerCase().includes(searchQuery.toLowerCase()) ||
-      book.author.toLowerCase().includes(searchQuery.toLowerCase()),
-  );
+  const booksPerPage = 30;
+
+  const fetchBooks = async (page: number) => {
```
Comment (GLObus303) [inline src/pages/Books/Books.tsx]: If you would do

```suggestion
  const fetchBooks = async (page: number = 1) => {
```

you don't need to pass in the 1 in the Effect and then you can do

```
  useEffect(() => {
    const timeoutId = setTimeout(fetchBooks, 500);

    return () => clearTimeout(timeoutId);
  }, [searchQuery]);
```

If you like it or not, I will leave it completely up to you! Just wanna show you the possibilities 🤩

File: src/pages/Books/Books.tsx [hunk 30]
```diff
@@ -1,51 +1,93 @@
...
+        ? await BookAPI.fetchSearchBooks({
+            query: searchQuery,
+            page,
+            limit: booksPerPage,
+          })
+        : await BookAPI.fetchTrendingBooks({
+            page,
+            limit: booksPerPage,
+          });
+
```
Comment (GLObus303) [inline src/pages/Books/Books.tsx:25]: I love this new line. It cures my OCD 😍

File: src/pages/BookDetail/BookDetail.tsx [hunk 31]
```diff
@@ -37,25 +27,131 @@ export const BookDetail: React.FC = () => {
...
+            ? new Date(bookData.first_publish_date).getFullYear()
+            : 'Unknown',
+          pages: bookData.number_of_pages || 'N/A',
+          rating: Math.random() * 5,
+          isCurrentlyReading: false,
+          editionKey: bookData.key,
+        };
+
+        setBook(bookDetails);
+        console.log('Book details:', bookDetails);
```
Comment (GLObus303) [inline src/pages/BookDetail/BookDetail.tsx]: 💢 I will pretend I haven't seen those other 17 logs above.

File: src/pages/BookDetail/BookDetail.tsx [hunk 32]
```diff
@@ -37,25 +27,131 @@ export const BookDetail: React.FC = () => {
...
+        if (!bookResponse.ok) {
+          throw new Error('Book not found');
+        }
+
+        const bookData = await bookResponse.json();
+
+        console.log('Fetched book data:', bookData);
+
+        let authorName = 'Unknown Author';
+        let authorId = '';
```
Comment (GLObus303) [inline src/pages/BookDetail/BookDetail.tsx]: First of all. This code is something I see every day. Trust me, people with years of experience consider this perfect. We go a step further 🚀 .

Let's avoid mutation. Especially this multiline one where multiple conditions are in play. This is what makes code veeery hard to follow and maintain

File: src/pages/BookDetail/BookDetail.tsx [hunk 33]
```diff
@@ -37,25 +27,131 @@ export const BookDetail: React.FC = () => {
...
+          ? `https://openlibrary.org/works/${queryId}.json`
+          : `https://openlibrary.org/books/${queryId}.json`;
+
+        const bookResponse = await fetch(endpoint);
+
+        if (!bookResponse.ok) {
+          throw new Error('Book not found');
+        }
+
+        const bookData = await bookResponse.json();
```
Comment (GLObus303) [inline src/pages/BookDetail/BookDetail.tsx]: Let's type this as I told you before with the explicit casting 

CleanShot 2024-12-08 at 23 51 30

We need types, everywhere. That is what will keep us from making silly mistakes.

I mean if we don't ignore it, riiiiiiiiiiiight
CleanShot 2024-12-08 at 23 52 02

File: src/pages/BookDetail/BookDetail.tsx [hunk 34]
```diff
@@ -37,25 +27,131 @@ export const BookDetail: React.FC = () => {
...
+
+          if (authorKey) {
+            authorId = parseAuthorId(authorKey);
+            const authorResponse = await fetch(
+              `https://openlibrary.org/authors/${authorId}.json`,
+            );
+            const authorData = await authorResponse.json();
+            authorName = authorData.name || authorName;
+          }
+        }
```
Comment (GLObus303) [inline src/pages/BookDetail/BookDetail.tsx]: These two blocks look very similar. I think you need one function "fetchDetail" that will take in the differences as params

File: src/pages/BookDetail/BookDetail.tsx [hunk 35]
```diff
@@ -37,25 +27,131 @@ export const BookDetail: React.FC = () => {
...
+        let authorName = 'Unknown Author';
+        let authorId = '';
+
+        if (isWorkKey) {
+          const authorKey = bookData.authors?.[0]?.author?.key;
+          console.log('Author Key:', authorKey);
+
+          if (authorKey) {
+            authorId = parseAuthorId(authorKey);
+            const authorResponse = await fetch(
```
Comment (GLObus303) [inline src/pages/BookDetail/BookDetail.tsx]: You have api layer (api folder), this should be there. No exception

File: src/pages/BookDetail/BookDetail.tsx [hunk 36]
```diff
@@ -37,25 +27,131 @@ export const BookDetail: React.FC = () => {
...
+            authorName = authorData.name || authorName;
+          }
+        }
+
+        const coverId = bookData.covers?.[0];
+        const coverUrl = coverId
+          ? `https://covers.openlibrary.org/b/id/${coverId}-L.jpg`
+          : '';
+
+        const bookDetails: Book = {
```
Comment (GLObus303) [inline src/pages/BookDetail/BookDetail.tsx]: We are mixing two separate processes together. That is not good -> https://react.dev/learn/lifecycle-of-reactive-effects#each-effect-represents-a-separate-synchronization-process

File: src/pages/BookDetail/BookDetail.tsx [hunk 37]
```diff
@@ -37,25 +27,131 @@ export const BookDetail: React.FC = () => {
...
+          cover: coverUrl,
+          description:
+            typeof bookData.description === 'string'
+              ? bookData.description
+              : bookData.description?.value || 'No description available',
+          year: bookData.first_publish_date
+            ? new Date(bookData.first_publish_date).getFullYear()
+            : 'Unknown',
+          pages: bookData.number_of_pages || 'N/A',
+          rating: Math.random() * 5,
```
Comment (GLObus303) [inline src/pages/BookDetail/BookDetail.tsx]: Are the books missing their rating whatsoever?

File: src/pages/BookDetail/BookDetail.tsx [hunk 38]
```diff
@@ -37,25 +27,131 @@ export const BookDetail: React.FC = () => {
...
+
+        const bookDetails: Book = {
+          id: queryId,
+          title: bookData.title || 'Untitled',
+          author: authorName,
+          cover: coverUrl,
+          description:
+            typeof bookData.description === 'string'
+              ? bookData.description
+              : bookData.description?.value || 'No description available',
```
Comment (GLObus303) [inline src/pages/BookDetail/BookDetail.tsx]: This is a pattern you might come across again. You can also solve it like this

```
   description: bookData.description?.value ?? bookData.description ?? 'No description available',
```

File: src/pages/BookDetail/BookDetail.tsx [hunk 39]
```diff
@@ -37,25 +27,131 @@ export const BookDetail: React.FC = () => {
...
+          `https://openlibrary.org/search.json?author=${encodeURIComponent(authorName)}&limit=5`,
+        );
+        const recommendedData = await recommendedResponse.json();
+        const recommendedBooksData: Book[] = recommendedData.docs
+          .filter((b: any) => b.key !== queryId)
+          .slice(0, 5)
+          .map((recommendedBook: any) => ({
+            id: parseAuthorId(recommendedBook.key),
+            title: recommendedBook.title,
+            author: recommendedBook.author_name || 'Unknown',
```
Comment (GLObus303) [inline src/pages/BookDetail/BookDetail.tsx]: I wonder if we should create like `defaultedValue` helper, since we have this bzilion of times. Not sure how it would look in terms of readability, I will leave for you to test what you prefer

**Review by GLObus303 — COMMENTED** [review]
Legit good work! I am impressed

File: .eslintrc [hunk 40]
```diff
@@ -35,7 +35,7 @@
   "rules": {
```
Comment (GLObus303) [inline .eslintrc:35]: CleanShot 2024-12-09 at 00 00 45

When you have 4 bullet points. Your PR is too big. Remember it next time.

File: src/pages/Books/Books.tsx (hunk 26 above)
Comment (eliskavo) [reply src/pages/Books/Books.tsx]: Good point 😄

File: .eslintrc (hunk 40 above)
Comment (GLObus303) [reply .eslintrc:35]: This is old comment, I forgot to send it. ☝️

File: .eslintrc [hunk 41]
```diff
@@ -35,7 +35,7 @@
...
     "strict": [2, "never"], // Avoid any unsafe functions that would not be allowed under strict mode
     "default-case": 0, // Allow switch without a default case
-    "no-console": 2, // Forbid console logs
+    "no-console": [2, { "allow": ["warn", "error"] }], // Forbid console logs
```
Comment (GLObus303) [inline .eslintrc]: Good job. But why warn? We dont have any warns

File: src/api/bookApi.tsx [hunk 42]
```diff
@@ -0,0 +1,87 @@
...
+  docs: Book[];
+  num_found: number;
+};
+
+type TrendingBooksResponse = {
+  works: Book[];
+};
+
+export const fetchSearchBooks = async ({ query }: FetchBooksParams) => {
+  const offset = (PAGE - 1) * LIMIT;
```
Comment (GLObus303) [inline src/api/bookApi.tsx]: This sounds like the offset will always be 0, isn't that correct?

File: src/api/bookApi.tsx [hunk 43]
```diff
@@ -0,0 +1,87 @@
...
+
+  const searchParams = new URLSearchParams({
+    q: query ?? '',
+    fields: 'key,title,author_name,cover_i,editions',
+    lang: 'eng,cze',
+    limit: LIMIT.toString(),
+    offset: offset.toString(),
+  });
+
+  const url = `https://openlibrary.org/search.json?${searchParams.toString()}`;
```
Comment (GLObus303) [inline src/api/bookApi.tsx]: Love it

File: src/api/bookApi.tsx [hunk 44]
```diff
@@ -0,0 +1,87 @@
...
+      total: data.num_found,
+    };
+  } catch (error) {
+    console.error('Error fetching search books:', error);
+
+    return { books: [], total: 0 };
+  }
+};
+
+export const fetchTrendingBooks = async () => {
```
Comment (GLObus303) [inline src/api/bookApi.tsx]: Note for call: Could these be abstracted(reused)

File: src/api/bookDetailApi.tsx [hunk 45]
```diff
@@ -0,0 +1,131 @@
...
+const fetchDetail = async (
+  endpoint: string,
+  isWorkKey: boolean,
+): Promise<Book | null> => {
+  try {
+    const bookResponse = await fetch(endpoint);
+    if (!bookResponse.ok) {
+      throw new Error('Book not found');
+    }
+    const bookData: any = (await bookResponse.json()) as Book;
```
Comment (GLObus303) [inline src/api/bookDetailApi.tsx]: Note for call: type

File: src/api/bookDetailApi.tsx [hunk 46]
```diff
@@ -0,0 +1,131 @@
...
+    console.error('Error fetching book rating:', error);
+
+    return 0;
+  }
+};
+
+const fetchDetail = async (
+  endpoint: string,
+  isWorkKey: boolean,
+): Promise<Book | null> => {
```
Comment (GLObus303) [inline src/api/bookDetailApi.tsx]: Note for call:  Perhaps no need for explicit return

File: src/api/bookDetailApi.tsx [hunk 47]
```diff
@@ -0,0 +1,131 @@
...
+    console.error('Error fetching book details:', error);
+
+    return null;
+  }
+};
+
+export const fetchBookDetails = async (
+  queryId: string,
+): Promise<Book | null> => {
+  const isWorkKey = queryId.startsWith('OL') && queryId.endsWith('W');
```
Comment (GLObus303) [inline src/api/bookDetailApi.tsx]: Note for call:  Could this be done differently

File: src/components/StarsRating/StarsRating.tsx [hunk 48]
```diff
@@ -0,0 +1,25 @@
+import style from './StarsRating.module.scss';
+
+type StarsProps = {
+  rating: number;
+  maxStars?: number;
+};
+
+export const Stars = ({ rating, maxStars = 5 }: StarsProps) => {
```
Comment (GLObus303) [inline src/components/StarsRating/StarsRating.tsx]: Note for call: We certainly should keep the same file name and component name in it.

Comment (GLObus303) [reply src/components/StarsRating/StarsRating.tsx]: Note for call:  Our typing convention is React.FC

File: src/components/StarsRating/StarsRating.tsx [hunk 49]
```diff
@@ -0,0 +1,25 @@
...
+  const roundedRating = Math.min(
+    Math.max(Math.round(rating * 2) / 2, 0),
+    maxStars,
+  );
```
Comment (GLObus303) [inline src/components/StarsRating/StarsRating.tsx:15]: This is some advanced math for me, I will need explanation :D

File: src/pages/BookDetail/BookDetail.tsx [hunk 50]
```diff
@@ -37,25 +31,62 @@ export const BookDetail: React.FC = () => {
...
+    const fetchRecommended = async () => {
+      if (!book?.author) {
+        return;
+      }
+
+      setIsLoadingRecommended(true);
+      try {
+        const recommendedBooksData = await fetchRecommendedBooks(
+          book.author,
+          queryId ?? '',
```
Comment (GLObus303) [inline src/pages/BookDetail/BookDetail.tsx]: Note for call:  I recommend passing object as params

File: src/pages/Books/Books.tsx [hunk 51]
```diff
@@ -54,14 +86,28 @@ export const Books: React.FC = () => {
...
+              </div>
+              <Pagination
+                currentPage={currentPage}
+                totalItems={totalItems}
+                itemsPerPage={BOOKS_PER_PAGE}
+                onPageChange={(page) => fetchBooks(page)}
+                showTotalPages
+              />
+            </>
+          )}
```
Comment (GLObus303) [inline src/pages/Books/Books.tsx]: Note for call: should these be components?

**Review by GLObus303 — APPROVED** [review]
